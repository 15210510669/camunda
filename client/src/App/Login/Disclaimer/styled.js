/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import styled from 'styled-components';
import {themed, themeStyle, Colors} from 'modules/theme';

const Container = themed(styled.div`
  color: ${themeStyle({
    dark: 'rgba(255, 255, 255, 0.7)',
    light: '#7e7e7f',
  })};
  opacity: 0.9;
  font-size: 12px;
  margin-top: 35px;
  width: 489px;
`);

const Anchor = themed(styled.a`
  text-decoration: underline;
  &:link {
    color: ${themeStyle({
      dark: Colors.darkLinkDefault,
      light: Colors.lightLinkDefault,
    })};
  }

  &:hover {
    color: ${themeStyle({
      dark: Colors.darkLinkHover,
      light: Colors.lightLinkHover,
    })};
  }

  &:active {
    color: ${themeStyle({
      dark: Colors.darkLinkActive,
      light: Colors.lightLinkActive,
    })};
  }

  &:visited {
    color: ${themeStyle({
      dark: Colors.darkLinkVisited,
      light: Colors.lightLinkVisited,
    })};
  }
`);

export {Container, Anchor};
