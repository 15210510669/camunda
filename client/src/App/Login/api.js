import {post, get} from 'modules/request';

export const login = async data => {
  const body = `username=${data.username}&password=${data.password}`;
  await post('/login', body, {
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded'
    }
  });
};

export const users = async userId => {
  const response = await get(`/users/${userId}`);
  const data = response.json();

  return data;
};
