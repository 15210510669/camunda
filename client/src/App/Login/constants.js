export const REQUIRED_FIELD_ERROR = 'Username and Password are both required';
export const LOGIN_ERROR = 'Username and Password do not match';
