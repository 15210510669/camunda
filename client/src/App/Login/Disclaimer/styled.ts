/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import styled, {css} from 'styled-components';

const Container = styled.div`
  ${({theme}) => {
    const colors = theme.colors.disclaimer.container;

    return css`
      color: ${colors.color};
      opacity: 0.9;
      font-size: 12px;
      margin-top: 35px;
      max-width: 489px;
    `;
  }}
`;

export {Container};
