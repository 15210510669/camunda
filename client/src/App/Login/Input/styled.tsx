/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import styled, {css} from 'styled-components';
import {styles} from '@carbon/elements';

const Container = styled.div`
  position: relative;
`;

const Input = styled.input`
  ${({theme}) => {
    const colors = theme.colors.login.input;

    return css`
      width: 100%;
      height: 48px;
      padding: 21px 8px 8px 8px;

      border-radius: 3px;
      border: solid 1px ${colors.borderColor};
      ${styles.bodyShort02};
      color: ${theme.colors.text01};

      background-color: ${colors.backgroundColor};

      & + label {
        pointer-events: none;
        user-select: none;
        position: absolute;
        top: 0px;
        left: 0px;

        transform: translate(9px, 5px);

        ${styles.label01};
        color: ${colors.labelColor};

        transition: all ease-in-out 150ms;
      }

      &:placeholder-shown:not(:focus-within) + label {
        transform: translate(9px, 14px);
        ${styles.label02};
        font-style: italic;
      }

      &:focus-visible {
        outline: none;
        box-shadow:
          0px 0px 0px 1px ${colors.focusInner},
          0px 0px 0px 4px ${theme.colors.focusOuter};
      }
    `;
  }}
`;

export {Container, Input};
