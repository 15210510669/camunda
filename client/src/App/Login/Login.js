/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import PropTypes from 'prop-types';
import {Redirect} from 'react-router-dom';

import {login} from 'modules/api/login';
import withSharedState from 'modules/components/withSharedState';
import Button from 'modules/components/Button';
import {PAGE_TITLE} from 'modules/constants';

import {REQUIRED_FIELD_ERROR, LOGIN_ERROR} from './constants';
import * as Styled from './styled';

class Login extends React.Component {
  static propTypes = {
    location: PropTypes.object.isRequired,
    clearStateLocally: PropTypes.func.isRequired
  };

  state = {
    username: '',
    password: '',
    forceRedirect: false,
    error: null
  };

  componentDidMount() {
    document.title = PAGE_TITLE.LOGIN;
  }

  handleLogin = async e => {
    e.preventDefault();
    const {username, password} = this.state;

    if (username.length === 0 || password.length === 0) {
      return this.setState({error: REQUIRED_FIELD_ERROR});
    }

    try {
      await login({username, password});
      this.props.clearStateLocally();
      this.setState({forceRedirect: true});
    } catch (e) {
      this.setState({error: LOGIN_ERROR});
    }
  };

  handleInputChange = ({target: {name, value}}) => {
    this.setState({[name]: value});
  };

  render() {
    const {username, password, forceRedirect, error} = this.state;

    // case of successful login
    if (forceRedirect) {
      const locationState = this.props.location.state || {referrer: '/'};
      return <Redirect to={locationState.referrer} />;
    }

    // default render
    return (
      <Styled.Login onSubmit={this.handleLogin}>
        <Styled.LoginHeader>
          <Styled.Logo />
          <Styled.LoginTitle>Operate</Styled.LoginTitle>
        </Styled.LoginHeader>
        <Styled.LoginForm>
          <Styled.FormError>{error}</Styled.FormError>
          <Styled.UsernameInput
            value={username}
            name="username"
            type="text"
            onChange={this.handleInputChange}
            placeholder="Username"
            aria-label="User Name"
          />
          <Styled.PasswordInput
            value={password}
            name="password"
            type="password"
            onChange={this.handleInputChange}
            placeholder="Password"
            aria-label="Password"
          />
          <Button type="submit" size="large" title="Log in">
            Log in
          </Button>
        </Styled.LoginForm>
        <Styled.Disclaimer>
          This Camunda Operate distribution is available under an evaluation
          license that is valid for development (non-production) use only. By
          continuing using this software, you agree to the{' '}
          <Styled.Anchor href="https://zeebe.io/legal/operate-evaluation-license">
            Terms and Conditions
          </Styled.Anchor>{' '}
          of the Operate Trial Version and our{' '}
          <Styled.Anchor href="https://zeebe.io/privacy/">
            Privacy Statement
          </Styled.Anchor>
          .
        </Styled.Disclaimer>
      </Styled.Login>
    );
  }
}

export default withSharedState(Login);
