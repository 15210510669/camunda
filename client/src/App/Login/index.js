/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import PropTypes from 'prop-types';
import {Redirect} from 'react-router-dom';

import {login} from 'modules/api/login';
import Button from 'modules/components/Button';
import {PAGE_TITLE} from 'modules/constants';
import {clearStateLocally} from 'modules/utils/localStorage';

import {Disclaimer} from './Disclaimer';
import {REQUIRED_FIELD_ERROR, LOGIN_ERROR} from './constants';
import * as Styled from './styled';
import SpinnerSkeleton from 'modules/components/SpinnerSkeleton';

class Login extends React.Component {
  static propTypes = {
    location: PropTypes.object.isRequired,
  };

  state = {
    username: '',
    password: '',
    forceRedirect: false,
    error: null,
    isLoading: false,
  };

  componentDidMount() {
    document.title = PAGE_TITLE.LOGIN;
  }

  handleLogin = async (e) => {
    e.preventDefault();
    const {username, password} = this.state;

    if (username.length === 0 || password.length === 0) {
      return this.setState({error: REQUIRED_FIELD_ERROR, isLoading: false});
    }

    this.setState({isLoading: true});

    try {
      await login({username, password});
      clearStateLocally();
      this.setState({forceRedirect: true, isLoading: false});
    } catch (e) {
      this.setState({error: LOGIN_ERROR, isLoading: false});
    }
  };

  handleInputChange = ({target: {name, value}}) => {
    this.setState({[name]: value});
  };

  render() {
    const {username, password, forceRedirect, error, isLoading} = this.state;

    // case of successful login
    if (forceRedirect) {
      const locationState = this.props.location.state || {referrer: '/'};
      return (
        <Redirect
          to={{
            pathname: locationState.referrer,
            state: {isLoggedIn: true},
          }}
        />
      );
    }

    // default render
    return (
      <>
        {isLoading && <SpinnerSkeleton data-test="spinner" />}
        <Styled.Login onSubmit={this.handleLogin}>
          <Styled.LoginHeader>
            <Styled.Logo />
            <Styled.LoginTitle>Operate</Styled.LoginTitle>
          </Styled.LoginHeader>
          <Styled.LoginForm>
            <Styled.FormError>{error}</Styled.FormError>
            <Styled.UsernameInput
              value={username}
              name="username"
              type="text"
              onChange={this.handleInputChange}
              placeholder="Username"
              aria-label="User Name"
            />
            <Styled.PasswordInput
              value={password}
              name="password"
              type="password"
              onChange={this.handleInputChange}
              placeholder="Password"
              aria-label="Password"
            />
            <Button
              data-test="login-button"
              type="submit"
              size="large"
              title="Log in"
            >
              Log in
            </Button>
          </Styled.LoginForm>
          <Disclaimer {...window.clientConfig} />
          <Styled.Copyright />
        </Styled.Login>
      </>
    );
  }
}

export {Login};
