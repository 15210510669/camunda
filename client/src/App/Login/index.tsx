/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useEffect} from 'react';
import {useHistory} from 'react-router-dom';
import {Form, Field} from 'react-final-form';
import {FORM_ERROR} from 'final-form';

import {login} from 'modules/api/login';
import Button from 'modules/components/Button';
import {PAGE_TITLE} from 'modules/constants';
import {clearStateLocally} from 'modules/utils/localStorage';

import {Disclaimer} from './Disclaimer';
import {LOGIN_ERROR, GENERIC_ERROR} from './constants';
import {
  Container,
  LoginHeader,
  Logo,
  LoginTitle,
  LoginForm,
  FormError,
  UsernameInput,
  PasswordInput,
  Copyright,
} from './styled';
import SpinnerSkeleton from 'modules/components/SpinnerSkeleton';
import {Routes} from 'modules/routes';

type FormValues = {
  username: string;
  password: string;
};

type LocationState = {
  referrer?: string;
  isLoggedIn?: boolean;
};

function Login() {
  const history = useHistory<LocationState>();

  useEffect(() => {
    document.title = PAGE_TITLE.LOGIN;
  }, []);

  return (
    <Form<FormValues>
      onSubmit={async (values) => {
        try {
          const response = await login(values);

          if (response.status === 401) {
            return {
              [FORM_ERROR]: LOGIN_ERROR,
            };
          }

          if (!response.ok) {
            return {
              [FORM_ERROR]: GENERIC_ERROR,
            };
          }
          clearStateLocally();
          history.push({
            pathname: history.location.state?.referrer ?? Routes.dashboard(),
            search: history.location.search,
            state: {
              isLoggedIn: true,
            },
          });
        } catch {
          return {
            [FORM_ERROR]: GENERIC_ERROR,
          };
        }
      }}
    >
      {({handleSubmit, submitting, submitError, form, dirtyFields}) => (
        <>
          {submitting && <SpinnerSkeleton data-testid="spinner" />}
          <Container>
            <LoginHeader>
              <Logo />
              <LoginTitle>Operate</LoginTitle>
            </LoginHeader>
            <LoginForm onSubmit={handleSubmit}>
              {submitError && <FormError>{submitError}</FormError>}
              <Field<FormValues['username']> name="username" component="input">
                {({input}) => (
                  <UsernameInput
                    name={input.name}
                    id={input.name}
                    value={input.value}
                    onChange={input.onChange}
                    type="text"
                    placeholder="Username"
                    aria-label="User Name"
                    required
                  />
                )}
              </Field>
              <Field<FormValues['password']> name="password" component="input">
                {({input}) => (
                  <PasswordInput
                    name={input.name}
                    id={input.name}
                    value={input.value}
                    onChange={input.onChange}
                    type="password"
                    placeholder="Password"
                    aria-label="Password"
                    required
                  />
                )}
              </Field>
              <Button
                type="submit"
                size="large"
                title="Log in"
                disabled={
                  !form
                    .getRegisteredFields()
                    .every((field) => dirtyFields[field]) || submitting
                }
              >
                Log in
              </Button>
            </LoginForm>
            <Disclaimer />
            <Copyright />
          </Container>
        </>
      )}
    </Form>
  );
}

export {Login};
