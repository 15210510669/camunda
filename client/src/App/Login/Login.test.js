import React from 'react';
import {shallow} from 'enzyme';
import {Redirect} from 'react-router-dom';

import {
  flushPromises,
  mockResolvedAsyncFn,
  mockRejectedAsyncFn
} from 'modules/testUtils';

import Login from './Login';
import * as Styled from './styled';
import * as api from 'modules/api/login';
import {REQUIRED_FIELD_ERROR, LOGIN_ERROR} from './constants';

jest.mock('modules/request');
jest.mock('modules/api/login');

describe('Login', () => {
  let node;
  let usernameInput;
  let passwordInput;

  beforeEach(() => {
    node = shallow(<Login location={{}} />)
      .first()
      .shallow();

    usernameInput = node.findWhere(
      element =>
        element.prop('type') === 'text' && element.prop('name') === 'username'
    );

    passwordInput = node.findWhere(
      element =>
        element.prop('type') === 'password' &&
        element.prop('name') === 'password'
    );
  });

  it('should render initially with no state data', () => {
    expect(node.state('username')).toEqual('');
    expect(node.state('password')).toEqual('');
    expect(node.state('forceRedirect')).toBe(false);
    expect(node.state('error')).toBeNull();
  });

  it('should render login form by default', () => {
    // given
    const submitInput = node.findWhere(
      element => element.prop('type') === 'submit'
    );

    // then
    expect(node.find(Styled.Login)).toHaveLength(1);
    expect(usernameInput).toHaveLength(1);
    expect(passwordInput).toHaveLength(1);
    expect(submitInput).toHaveLength(1);
    expect(node).toMatchSnapshot();
  });

  it('should change state according to inputs change', () => {
    //given
    const username = 'foo';
    const password = 'bar';

    // when
    usernameInput.simulate('change', {
      target: {name: 'username', value: username}
    });

    passwordInput.simulate('change', {
      target: {name: 'password', value: password}
    });

    // then
    expect(node.state('username')).toEqual(username);
    expect(node.state('password')).toEqual(password);
  });

  it('should reset the stored state on login', async () => {
    // given
    const spy = jest.fn();
    node.setProps({clearStateLocally: spy});
    node.setState({username: 'foo', password: 'bar'});

    // when
    node.instance().handleLogin({preventDefault: () => {}});
    await flushPromises();

    // then
    expect(spy).toHaveBeenCalled();
  });

  describe('redirection', () => {
    let originalLogin = api.login;
    let username = 'foo',
      password = 'bar';
    api.login = mockResolvedAsyncFn();

    beforeEach(() => {
      api.login.mockClear();
      node.setState({username, password});
    });

    afterAll(() => {
      // reset api.login
      api.login = originalLogin.bind(api);
    });

    it('should redirect to home page on successful login', async () => {
      // when
      node.instance().handleLogin({preventDefault: () => {}});
      await flushPromises();
      node.update();

      // then
      expect(api.login).toBeCalledWith({username, password});
      const RedirectNode = node.find(Redirect);
      expect(RedirectNode).toHaveLength(1);
      expect(RedirectNode.prop('to')).toBe('/');
      expect(node).toMatchSnapshot();
    });

    it('should redirect to referrer page on successful login', async () => {
      // given
      const referrer = '/some/page';
      node.setProps({location: {state: {referrer}}});

      // when
      node.instance().handleLogin({preventDefault: () => {}});
      await flushPromises();
      node.update();

      // then
      expect(api.login).toBeCalledWith({username, password});
      const RedirectNode = node.find(Redirect);
      expect(RedirectNode).toHaveLength(1);
      expect(RedirectNode.prop('to')).toBe(referrer);
      expect(node).toMatchSnapshot();
    });
  });

  it('should display an error if any field is empty', async () => {
    // mock api.login
    const originalLogin = api.login;
    api.login = jest.fn();

    // when
    node.instance().handleLogin({preventDefault: () => {}});
    await flushPromises();
    node.update();

    // then
    expect(api.login).not.toHaveBeenCalled();
    const errorSpan = node.find(Styled.FormError).render();
    expect(node.state('error')).toEqual(REQUIRED_FIELD_ERROR);
    expect(errorSpan.text()).toContain(REQUIRED_FIELD_ERROR);
    expect(node).toMatchSnapshot();

    // reset api.login
    api.login = originalLogin.bind(api);
  });

  it('should display an error on unsuccessful login', async () => {
    // mock api.login
    const originalLogin = api.login;
    api.login = mockRejectedAsyncFn();
    node.setState({username: 'foo', password: 'bar'});

    // when
    node.instance().handleLogin({preventDefault: () => {}});
    await flushPromises();
    node.update();

    // then
    const errorSpan = node.find(Styled.FormError).render();
    expect(node.state('error')).toEqual(LOGIN_ERROR);
    expect(errorSpan.text()).toContain(LOGIN_ERROR);
    expect(node).toMatchSnapshot();

    // reset api.login
    api.login = originalLogin.bind(api);
  });
});
