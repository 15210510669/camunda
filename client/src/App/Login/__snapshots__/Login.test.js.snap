// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login redirection should redirect to home page on successful login 1`] = `
<Redirect
  push={false}
  to="/"
/>
`;

exports[`Login redirection should redirect to referrer page on successful login 1`] = `
<Redirect
  push={false}
  to="/some/page"
/>
`;

exports[`Login should display an error if any field is empty 1`] = `
<styled.div
  onSubmit={[Function]}
>
  <Themed(styled.div)>
    <Themed(Styled(ReactComponent)) />
    <Themed(styled.span)>
      Operate
    </Themed(styled.span)>
  </Themed(styled.div)>
  <styled.form>
    <styled.span>
      Username and Password are both required
    </styled.span>
    <Themed(Styled(Styled(TextInput)))
      aria-label="User Name"
      name="username"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value=""
    />
    <Themed(Styled(Styled(TextInput)))
      aria-label="Password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <Button
      size="large"
      title="Log in"
      type="submit"
    >
      Log in
    </Button>
  </styled.form>
</styled.div>
`;

exports[`Login should display an error on unsuccessful login 1`] = `
<styled.div
  onSubmit={[Function]}
>
  <Themed(styled.div)>
    <Themed(Styled(ReactComponent)) />
    <Themed(styled.span)>
      Operate
    </Themed(styled.span)>
  </Themed(styled.div)>
  <styled.form>
    <styled.span>
      Username and Password do not match
    </styled.span>
    <Themed(Styled(Styled(TextInput)))
      aria-label="User Name"
      name="username"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value="foo"
    />
    <Themed(Styled(Styled(TextInput)))
      aria-label="Password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="bar"
    />
    <Button
      size="large"
      title="Log in"
      type="submit"
    >
      Log in
    </Button>
  </styled.form>
</styled.div>
`;

exports[`Login should render login form by default 1`] = `
<styled.div
  onSubmit={[Function]}
>
  <Themed(styled.div)>
    <Themed(Styled(ReactComponent)) />
    <Themed(styled.span)>
      Operate
    </Themed(styled.span)>
  </Themed(styled.div)>
  <styled.form>
    <Themed(Styled(Styled(TextInput)))
      aria-label="User Name"
      name="username"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value=""
    />
    <Themed(Styled(Styled(TextInput)))
      aria-label="Password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <Button
      size="large"
      title="Log in"
      type="submit"
    >
      Log in
    </Button>
  </styled.form>
</styled.div>
`;
